### ISSUE TITLE
WISH: way to temporarily disable GC or at least read its state

### ISSUE THREAD

**Issue by hiiamboris** [red/REP#130]
We have `recycle/on` and `recycle/off`, but what if I want to disable GC and then restore its previous state (unknown to me)?

I have to currently resort to this workaround:
`without-GC: func [code] [sort/compare [1 1] func [a b] code]`

Would be nice to have a by-design way to do that.

**Comment by greggirwin** [comment]
This is a good general design question for any global state. The problem being how to know if the state changed since you last checked it, so you don't restore a state that messes up other code.

**Comment by hiiamboris** [comment]
If you only change state for the code under your control, and restore it upon leaving this code, it should be fine.

**Comment by dockimbel** [comment]
>  restore its previous state

The GC does not have "states" outside of the moments it's running. There is no "previous" GC state. Garbage collecting is a process, not a data-structure. Please reformulate and elaborate on what you mean precisely.

> I have to currently resort to this workaround:
> without-GC: func [code] [sort/compare [1 1] func [a b] code]

I have no clue what you want to achieve their nor why. Instead of playing once again the guessing game, please be exhaustive and precise in your descriptions.

**Comment by hiiamboris** [comment]
It has `on` and `off` states (collector/active?). Come on.
```
REFINEMENTS:
     /on          => Turns on garbage collector; returns nothing.
     /off         => Turns off garbage collector; returns nothing.
```

**Comment by hiiamboris** [comment]
Ping?
My `without-GC` trick got broken somewhere between Mar 2 and Mar 7 of 2025. 
Since then there's no way to temporarily disable GC on Red level left.

**Comment by dockimbel** [comment]
We could add a `/disable` refinement that would take an additional body block argument where GC will be disabled. Exiting the body block will return the GC to its previous state. Would that be enough?

**Comment by hiiamboris** [comment]
Yes that'll be fine. Just need a name. `/disable` is kinda synonymous with `/off` and doesn't imply temporary nature of it. Maybe `recycle/pause` or `recycle/hold`?

**Comment by greggirwin** [comment]
Does it make more sense to have a `do` related helper, like we have for `do-safe` and `do-no-sync`? This feels more like an env/call-level control feature to me.

**Comment by greggirwin** [comment]
Could also be a refinement on `do`, as it's akin to `/expand` and `/trace` .

**Comment by hiiamboris** [comment]
Finding a meaningful name for collection negation with `do` is challenging 😔

**Comment by greggirwin** [comment]
Naming always is. Question 1 for me is what makes the most sense semantically. I agree that `/disable` sounds just like `/off` in that regard, and putting this behavior in `recycle` seems backwards, and needs a better name to indicate that an extra param, the block, to go with it.

GC is an advanced concept, `recycle` is friendlier, `do/no-recycle` is a bit long, but `do/no-GC` introduces a new term.

**Comment by dockimbel** [comment]
There are many features that could be applied locally only, adding new refinements for that in `do` is not scalable.

What about rather exposing the recycling state and use a mezz wrapper instead?

E.g., extend `recycle/info` to also return a logic value to reflect the active internal state? Though, that mean returning a block, so one extra very short-lived series allocation... Maybe a specific refinement then, like `recycle/active?`.

**Comment by greggirwin** [comment]
I agree that a wrapper of some kind, for local environment control is the best approach. Maybe we can think about a systemic state capture of some kind, for this kind of control.

**Comment by hiiamboris** [comment]
To expose state or only a refinement depends on whether:
- we expect to need to temporarily turn GC on while keeping it generally off (I don't think so, since we have `recycle` manual call)
- we expect some Red-level tools to read that state, maybe for debugging, reporting or some optimizations (which I guess is a possibility)

**Comment by dockimbel** [comment]
@hiiamboris Second option.

>  Maybe we can think about a systemic state capture of some kind, for this kind of control.

We have `system/state` object for exposing some internal states to the user. Though, exposing a R/S level state requires a reactive object, so, performance-wise, not best suited for intensive querying. Using a _pull_ approach there rather than a _push_ one would be way simpler and faster, but also an adhoc solution that would multiply functions or refinements as  more states are exposed.

Maybe we can start with the best option architecture-wise (extending `system/state`), then see if it becomes a bottleneck or not?

**Comment by hiiamboris** [comment]
You mean, we replace `collector/active?: ...` with a function, like `collector/set-active ...` and it will set a `system/state/collecting?` (or similar) flag in sync with changes? 

Works for me, and the design maturity requirements for `system/state` are lower than for natives.

**Comment by dockimbel** [comment]
> You mean, we replace collector/active?: ... with a function, like collector/set-active ... and it will set a system/state/collecting? (or similar) flag in sync with changes?

Could be done that way (info pushing), but just reading that value when user would access `system/state/gc-active?` would be enough (info on-demand pulling).

I'll make a simple implementation accordingly.

**Comment by dockimbel** [comment]
I forgot we still don't have `get` object event, so the only way is using routines in `system/state`. It's very efficient, but prevents from PROBE-ing the object and getting the values directly...

**Comment by dockimbel** [comment]
Added in https://github.com/red/red/commit/5399e7e83798a2a207a04d62422bcc75ffc276c8

**Comment by hiiamboris** [comment]
Obsoletes `recycle/info`?

**Comment by dockimbel** [comment]
Yes, we can remove that refinement.
